tables/Tables/BaseTable.cc

```diff
@@ -745,32 +726,42 @@ BaseTable* BaseTable::doSort (PtrBlock<BaseColumn*>& sortCol,
     //# Create a reference table.
     //# This table will NOT be in row order.
     rownr_t nrrow = nrow();
-    RefTable* resultTable = makeRefTable (False, nrrow);
+    std::shared_ptr<BaseTable> resultBaseTab = makeRefTable (False, nrrow);
+    RefTable* resultTable = resultBaseTab->asRefTable();
     //# Now sort the table storing the row-numbers in the RefTable.
     //# Adjust rownrs in case source table is already a RefTable.
     //# Then delete possible allocated data blocks.
-    Vector<rownr_t>& rows = *(resultTable->rowStorage());
+    Vector<rownr_t>& rows = resultTable->rowStorage();
     //# Note that nrrow can change in case Sort::NoDuplicates was given.
     nrrow = sortobj.sort (rows, nrrow, option);
-    if(sortIterBoundaries && sortIterKeyIdxChange)
+    if(sortIterBoundaries && sortIterKeyIdxChange) {
         sortobj.unique(*sortIterBoundaries, *sortIterKeyIdxChange, rows);
+    }
     adjustRownrs (nrrow, rows, False);
     resultTable->setNrrow (nrrow);
-    return resultTable;
+    return resultBaseTab;
 }
 
-RefTable* BaseTable::makeRefTable (Bool rowOrder, rownr_t initialNrrow)
+std::shared_ptr<BaseTable> BaseTable::makeRefTable (Bool rowOrder,
+                                                    rownr_t initialNrrow)
 {
-    RefTable* rtp = new RefTable(this, rowOrder, initialNrrow);
-    return rtp;
+  std::shared_ptr<BaseTable> rtp (new RefTable(this, rowOrder, initialNrrow));
+  rtp->setWeakPtr (rtp);
+  return rtp;
 }
 
+RefTable* BaseTable::asRefTable()
+{
+  RefTable* rtp = dynamic_cast<RefTable*>(this);
+  AlwaysAssert (rtp, AipsError);
```

Comment:
Why an AlwaysAssert? Generally programming errors are checked using a debug assert that is compiled out when doing a production compilation, whereas possible users (e.g. error in parset) are more strictly checked.

(I'm also not sure this function is too desirable. It might be better to dynamic_cast where necessary instead of adding an extra indirection for the programmer, that now has to look up what asRefTable does.)

Reply:
Removed asRefTable and used DebugAssert